idaes/apps/uncertainty_propagation/uncertainties.py

```diff
@@ -0,0 +1,271 @@
+    if not solver_options==None:
+        if not isinstance(solver_options, dict):
+            raise TypeError('solver_options must be dictionary.')
+    # Remove all "'" and " " in theta_names
+    theta_names, var_dic,variable_clean = clean_variable_name(theta_names)
+    parmest_class = parmest.Estimator(model_function, data, theta_names, obj_function,
+                 tee, diagnostic_mode, solver_options)
+    obj, theta,cov = parmest_class.theta_est(calc_cov=True)
+    # Convert theta keys to the original name 
+    # Revert theta_names to be original
+    if variable_clean:
+        theta_out = {}
+        for i in var_dic.keys():
+            theta_out[var_dic[i]] = theta[i]
+        theta_names = [var_dic[v] for v in theta_names] 
+    else:
+        theta_out = theta
+    propagate_results  =  propagate_uncertainty(model_uncertain, theta, cov, theta_names, tee)
+   
+    Output = namedtuple('Output',['obj', 'theta', 'theta_names', 'cov','gradient_f', 'gradient_c', 'dsdp', 'propagation_c', 'propagation_f','col','row'])
+    results= Output(obj, theta_out, theta_names, cov, 
+                     propagate_results.gradient_f,
+                     propagate_results.gradient_c, 
+                     propagate_results.dsdp,
+                     propagate_results.propagation_c,
+                     propagate_results.propagation_f,
+                     propagate_results.col,
+                     propagate_results.row
+                     )
+    return results
+
+def propagate_uncertainty(model_uncertain, theta, cov, theta_names, tee=False, solver_options=None):
+    """
+    This function calculates gradient vector, expectation, and variance of the objective function and constraints  of the model for given estimated optimal parameters and covariance matrix of parameters. It calculates error propagation of the objective function and constraints by using gradient vector and covariance matrix. 
+    The following terms are used to define the output dimensions:
+    Ncon   = number of constraints
+    Nvar   = number of variables (Nx + Ntheta)
+    Nx     = the numer of decision (primal) variables
+    Np = number of uncertain parameters.
+
+    Args:
+        model_uncertain(function or Pyomo ConcreteModel) : Function is a a python Function that generates an instance of the Pyomo model
+        theta(dict)                    : Size Ntheta python dictionary. Estimated parameters
+        cov(numpy.ndarray)             : Ntheta by Ntheta matrix. Covariance matrix of parameters
+        theta_names(list of strings)   : Size Ntheta. List of estimated l theta names
+        var_dic(dictionary)            : Size Ntheta. If any original variable contains apostrophes, need an auxiliary dictionary with keys theta_namess without apostrophes, values with apostrophes. e.g) var_dic: {'fs.properties.tau[benzene,toluene]': "fs.properties.tau['benzene','toluene']", 'fs.properties.tau[toluene,benzene]': "fs.properties.tau['toluene','benzene']"} 
```

Comment:
I don't see var_dic in this function's arguments. Should it be there?

Reply:
you are right. Removed